---
title: "ECL_Zombie_CuffDiff"
author: "Susan Moenga"
date: "2/26/2018"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r CUFFDIFF}
library(cummeRbund)
```



```{r pressure, echo=FALSE}
plot(pressure)

#read in the data 
setwd("/Volumes/smmoenga/ECL243/Zombie_data/cuffdiff_out")
cuff<-readCufflinks()
cuff



#add annotation info
setwd("/Volumes/smmoenga/ECL243/Zombie_data/data")
annot <-read.delim("GCA_001272575.2_ASM127257v2_genomic.txt", sep="\t",header=T,na.string="-")

addFeatures(cuff,annot,level="genes")

#####Global statistics and quality control##
#dispersion
disp<-dispersionPlot(genes(cuff))
disp


#variation

genes.scv<-fpkmSCVPlot(genes(cuff))
genes.scv
isoforms.scv<-fpkmSCVPlot(isoforms(cuff))
isoforms.scv

#To assess the distributions of FPKM scores across samples, you can use the csDensity plot

##all summarized
dens<-csDensity(genes(cuff))
dens

##with replicates indepently shown
densRep<-csDensity(genes(cuff),replicates=T)
densRep

brep<-csBoxplot(genes(cuff),replicates=F)
 brep
 
 ##global clustering?
dend.rep<-csDendro(genes(cuff),replicates=T)
dend.rep


##MvsA plots for detecting systematic bias

m1<-MAplot(genes(cuff),"Live_manipulation","Fungal_Control")
m1

m2<-MAplot(genes(cuff),"Dead_after_Mani","Fungal_Control")
m2


## CONDITIONAL COMPARISONS
v<-csVolcanoMatrix(genes(cuff))
v


#For individual pairwise comparisons, you must specify the comparisons by sample name.
v1 = csVolcano(genes(cuff), "Live_manipulation","Fungal_Control", alpha=0.05, showSignificance = T)
v1
v2 = csVolcano(genes(cuff), "Dead_after_Mani","Fungal_Control") # Dead vs control

v3 = csVolcano(genes(cuff), "Live_manipulation","Fungal_Control")
v3

#get run info on cuffdiff 
runInfo(cuff)  

#######subsetting, whatever feature that you want####


fpkm_4_filtered=subset(fpkm_all,(fpkm >4))
head(fpkm_4_filtered)
nrow(fpkm_4_filtered)


###use this list to call out important/filtered genes in cuff
filtered_genes = fpkm_4_filtered[ , "gene_id"]


#To facilitate Bioconductor-like operations, an 'FPKM-matrix' can be returned easily using the fpkmMatrix method:
gene.matrix<-fpkmMatrix(genes(cuff)) # without reps
head(gene.matrix)

gene.rep.matrix<-repFpkmMatrix(genes(cuff))  #with reps
head(gene.rep.matrix)


#Similarly, a matrix of normalized counts can be generated by using countMatrix
gene.count.matrix<-countMatrix(genes(cuff))
head(gene.count.matrix)



#####Creating gene sets from significantly regulated genes

 ## all conditions 
mySigGeneIds<-getSig(cuff,alpha=0.05,level='genes')
head(mySigGeneIds)

##how many?
length(mySigGeneIds)


## specific DEG conditions
#Live_Control = 3264 genes
Live_vs_Ctrl.sigGeneIds<-getSig(cuff,'Live_manipulation','Fungal_Control',alpha=0.05,level='genes')
length(Live_vs_Ctrl.sigGeneIds)
head(Live_vs_Ctrl.sigGeneIds)



#Dead_Control = 2817
Dead_vs_Ctrl.sigGeneIds<-getSig(cuff,x='Dead_after_Mani',y='Fungal_Control',alpha=0.05,level='genes')
length(Dead_vs_Ctrl.sigGeneIds)
head(Dead_vs_Ctrl.sigGeneIds)


#### ESSENTIALLY START HERE ###### 

###### Significantly differentially expressed
gene_diff_data <- diffData(genes(cuff))
sig_filtered_data <- subset(gene_diff_data, (value_1>4)) #remove any below 4 fpkm count
sig_filtered_data1 <- subset(sig_filtered_data, (value_2>4))#remove any below 4 fpkm count
sig_gene_data  <- subset(sig_filtered_data1, (significant ==  'yes')) 


#live_ctr subset 
sig_gene_data_live = subset(sig_gene_data, (sample_1 == "Live_manipulation"))
sig_gene_data_live_control=subset(sig_gene_data_live, (sample_2=="Fungal_Control"))

nrow(sig_gene_data_live_control) ##how many?  # 2232

up_live_ctrl  <- subset(sig_gene_data_live_control, (log2_fold_change > 1))
nrow(up_live_ctrl) # 1235, they found 1417

down_live_ctrl<- subset(sig_gene_data_live_control, (log2_fold_change < -1))
nrow(down_live_ctrl)  #941



 ##Dead_after manipulation_ctrl
 
 sig_gene_data_dead = subset(sig_gene_data, (sample_1 == "Dead_after_Mani"))
sig_gene_data_dead_control=subset(sig_gene_data_dead, (sample_2=="Fungal_Control"))

nrow(sig_gene_data_dead_control) ##how many?  # 1932

up_dead_ctrl  <- subset(sig_gene_data_dead_control, (log2_fold_change > 1))
nrow(up_dead_ctrl) # 942, they found 1050

down_dead_ctrl<- subset(sig_gene_data_dead_control, (log2_fold_change < -1))
nrow(down_dead_ctrl)  #602

##investigate what was up then down after

up_before1 = subset(gene_diff_data, (sample_1=="Live_manipulation"))
up_before2 = subset(up_before1, (sample_2=="Fungal_Control"))
up_before = subset(up_before2, (log2_fold_change <0 ))
up_down_list = up_before[, "gene_id"]
length(up_down_list)
down_after = getGenes(cuff,up_down_list, sampleIdList = c("Dead_after_Mani", "Fungal_Control")) #cufflist 

up_then_down_after_Mani = subset(diffData(down_after), (log2_fold_change > 0))
up_then_down_after_Mani1 = subset(up_then_down_after_Mani,(significant == "yes"))
nrow(up_then_down_after_Mani1)


#now do down on the above cufflist

str(down_dead_ctrl)

#####or get them in a signifance matrix

mySigTable<-getSigTable(cuff,alpha=0.01,level='genes')  #can adjust alpha to whatever
head(mySigTable,20)


```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.



```{r  Exploring the relationships between conditions}


#Distance matrix
myDistHeat<-csDistHeat(genes(cuff))
myDistHeat

#with reps
myRepDistHeat<-csDistHeat(genes(cuff),replicates=T)
myRepDistHeat


## Dimensionality reduction

genes.PCA<-PCAplot(genes(cuff),"PC1","PC2", replicates = T)
genes.PCA

genes.MDS<-MDSplot(genes(cuff), replicates = T)
genes.MDS


```
```{r  Partitioning}

```


```{r Ant alignmnet}
setwd("/Volumes/smmoenga/ECL243/Zombie_data/cuffdiff_ant_out")
cuff_ant <- readCufflinks()


v_ant<-csVolcanoMatrix(genes(cuff_ant))
v_ant


gene_diff_data_ant <- diffData(genes(cuff_ant))
sig_filtered_data_ant <- subset(gene_diff_data_ant, (value_1>4)) #remove any below 4 fpkm count
sig_filtered_data1_ant <- subset(sig_filtered_data_ant, (value_2>4))#remove any below 4 fpkm count
sig_gene_data_ant  <- subset(sig_filtered_data1_ant, (significant ==  'yes')) 

#specific conditions 

##1 - Live vs 10 am
sig_gene_data_live_ant = subset(sig_gene_data_ant, (sample_1 == "Live_manipulation"))
sig_gene_data_live_control_ant=subset(sig_gene_data_live_ant, (sample_2=="Ant_Control_10am"))

nrow(sig_gene_data_live_control_ant) ##how many?  # 1386

up_live_ctrl_ant  <- subset(sig_gene_data_live_control_ant, (log2_fold_change > 1))
nrow(up_live_ctrl_ant) # 439,

down_live_ctrl_ant<- subset(sig_gene_data_live_control_ant, (log2_fold_change < -1))
nrow(down_live_ctrl_ant)  #265


##2 - Dead vs 10 am

sig_gene_data_dead_ant = subset(sig_gene_data_ant, (sample_2 == "Dead_after_Mani"))
sig_gene_data_dead_control_ant=subset(sig_gene_data_dead_ant, (sample_1=="Ant_Control_10am"))

nrow(sig_gene_data_live_control_ant) ##how many?  # 1386

up_live_ctrl_ant  <- subset(sig_gene_data_live_control_ant, (log2_fold_change > 1))
nrow(up_live_ctrl_ant) # 439,

down_live_ctrl_ant<- subset(sig_gene_data_live_control_ant, (log2_fold_change < -1))
nrow(down_live_ctrl_ant)  #265


```




